### PR TITLE
Initial commit.

### PR DESCRIPTION
**Pull request by erikcorry**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,54 @@
...
+## Differences to xxd
+
+* Long options must be used with double dashes, not single dashes.
+* The single letter options are the same as xxd, but some long form options have been added:
+    * `--compact` A more compact output format for C format.
+    * `--indentation` Change the default indentation (2 for C format, 0 for other formats).
+    * `--little-endian` An option that changes the endianness of multi-byte words without changing the format.
+    * `--tabs` Use a mixture of tabs and spaces for indentation.
+    * `--tab-width` Change the default tab width if tabs are used (8).
+* Does not have a -r option to convert text back to binary. The `--seek` option is therefore also not supported.
```
Comment (kasperl) [inline README.md:20]:
```suggestion
* Does not have a `-r` option to convert text back to binary. The `--seek` option is therefore also not supported.
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,54 @@
...
+* The `-c` (columns) option is not limited to 256.
+* The `-E` (EBCDIC) option is not supported.
+* The `-p` (postscript) option has only one long form option, `--postscript`.  The `--ps` and `--plain` aliases are not supported.
+* The `-s` option does not accept unary plus, and negative seeks are not supported.
+
+# Usage
+
+```
+Usage:
+  xxxd.toit [<options>] [--] [<in:file>] [<out:file>]
```
Comment (kasperl) [inline README.md:30]: I imagine we will ship this as precompiled binaries so it probably should just be `xxxd`.

File: app/xxxd.toit [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+            --default=""
+            --short-help="Output (default stdin)"
+            --type="file",
+        ]
+    --run= :: convert it
+  root-cmd.run args
+
+convert parsed -> none:
+  if parsed["version"]:
+    print "xxxd in Toit v0.1.0"
```
Comment (kasperl) [inline app/xxxd.toit:107]: Is there a non-Toit version of xxxd?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,54 @@
 # xxxd
-An alternative to xxd - binary to source converter - download binaries for major OSs from the releases page.
+An alternative to xxd.  Xxd is sometimes distributed with vim, whereas this is standalone.
```
Comment (floitsch) [inline README.md:2]: Markdown wants a new line after sections. (By convention, not by necessity).
```suggestion

An alternative to xxd.  Xxd is sometimes distributed with vim, whereas this is standalone.
```

File: app/xxxd.toit [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+
+import binary show LITTLE_ENDIAN BIG_ENDIAN
+import cli
+import host.pipe
+import host.file
+import reader show BufferedReader
+
+main args:
+  root-cmd := cli.Command "convert"
+    --long-help="""
```
Comment (floitsch) [inline app/xxxd.toit:13]: Typically we indent parameters by 4.

File: app/xxxd.toit [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+        Convert a binary file to a source code file.
+        Options are largely compatible with the xxd utility from vim.
+        The 'r' (reverse) command is not supported.
+        Single-dash long options like '-name' are not supported: Use '--name' instead.
```
Comment (floitsch) [inline app/xxxd.toit:17]: I would drop this line. It's mentioned in the readme.

File: app/xxxd.toit [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+  c.dump parsed
+
+class Convert:
+  line-start/LineStart
+  word-formatter/WordFormatter
+  separator/Separator
+  trailer/Trailer
+
+  auto-skip/bool
+  declaration-allowed := false
```
Comment (floitsch) [inline app/xxxd.toit:120]:
```suggestion
  declaration-allowed/bool := false
```

File: app/xxxd.toit [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+          ba[it] = '_'
+      name = ba.to-string
+    if name != "" and '0' <= name[0] <= '9':
+      name = "__$name"
+    if parsed["capitalize"]:
+      name = name.to-ascii-upper
+
+    writer := out-name == "" ?
+      pipe.stdout :
+      file.Stream.for-write out-name
```
Comment (floitsch) [inline app/xxxd.toit:219]:
```suggestion
    writer := out-name == ""
        ? pipe.stdout
        : file.Stream.for-write out-name
```

File: app/xxxd.toit [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+    in-all-zero := false
+    all-zero-printed := false
+
+    seek-input := parsed["seek-input"]
+    if seek-input < 0: throw "-s options must be non-negative"
+    reader.skip seek-input
+    pos := seek-input
+
+    while reader.can-ensure 1:
+      bytes := ?
```
Comment (floitsch) [inline app/xxxd.toit:233]:
```suggestion
      bytes/int := ?
```

File: app/xxxd.toit [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+abstract class FormatFormatter implements WordFormatter:
+  format-string/string? := null
+  format-letter/string := ?
+  chars-per-byte/int := ?
+  group-size/int := 0
+
+  constructor .format-letter .chars-per-byte:
+
+  format word/int --chunk/int -> string:
+    fmt := ?
```
Comment (floitsch) [inline app/xxxd.toit:357]:
```suggestion
    fmt/string := ?
```

**Review by floitsch — APPROVED** [review]
LGTM.

**Pull request merged**